src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+
+
+@dataclass(frozen=True)
+class SetupPySourcesRequest:
+  hydrated_targets: HydratedTargets
+
+
+@dataclass(frozen=True)
+class SetupPySources:
+  """The sources required by a setup.py command.
+
+  Includes some information derived from analyzing the source, namely the packages,
+  namespace packages and resource files in the source.
+  """
+  digest: Digest
+  packages: Tuple[str, ...]
+  namespace_packages: Tuple[str, ...]
+  package_data: Tuple[PackageDatum, ...]
+
+
+@dataclass(frozen=True)
+class SetupPyChrootRequest:
+  """A request to create a chroot containing a setup.py and the sources it operates on."""
+  exported_target: ExportedTarget
+  py2: bool = False  # Whether to use py2 or py3 namespace package semantics.
+
+
+@dataclass(frozen=True)
+class SetupPyChroot:
+  """A chroot containing a generated setup.py and the sources it operates on."""
+  digest: Digest
+  # The keywords are embedded in the setup.py file in the digest, so these aren't
+  # strictly needed here, but they are convenient for testing.
+  setup_keywords_json: str
+
+
+@dataclass(frozen=True)
+class RunSetupPyRequest:
+  """A request to run a setup.py command."""
+  exported_target: ExportedTarget
+  chroot: SetupPyChroot
+  args: Tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class RunSetupPyResult:
+  """The result of running a setup.py command."""
+  output: Digest  # The state of the chroot after running setup.py.
+
+
+@dataclass(frozen=True)
+class SetuptoolsSetup:
+  """The setuptools tool."""
+  requirements_pex: Pex
+
+
+class SetupPyOptions(GoalSubsystem):
+  name = "setup-py2"
+
+  @classmethod
+  def register_options(cls, register):
+    super().register_options(register)
+    register(
+      '--args', type=list, member_type=str,
+      help="Arguments to pass directly to setup.py, e.g. "
+           "`--setup-py2-args=\"bdist_wheel\"`. If unspecified, we just dump the setup.py chroot."
+    )
+
+
+class SetupPy(Goal):
+  """Runs setup.py commands."""
+  subsystem_cls = SetupPyOptions
+
+
+@console_rule
+async def run_setup_pys(addresses: BuildFileAddresses, console: Console, options: SetupPyOptions,
+                        distdir: DistDir, workspace: Workspace) -> SetupPy:
+  """Run setup.py commands on all exported targets addressed."""
+  targets = await Get[HydratedTargets](BuildFileAddresses, addresses)
+
+  exported_targets: List[ExportedTarget] = []
+  nonexported_targets: List[HydratedTarget] = []
+  for target in targets:
+    if _is_exported(target):
+      exported_targets.append(ExportedTarget(target))
+    else:
+      nonexported_targets.append(target)
+  if nonexported_targets:
+    raise TargetNotExported(
+      'Cannot run setup.py on these targets, because they have no `provides=` clause: '
+      f'{", ".join(so.address.reference() for so in nonexported_targets)}')
+
+  for target in exported_targets:
+    addr = target.hydrated_target.address.reference()
+    chroot = await Get[SetupPyChroot](SetupPyChrootRequest(target))
```

Comment:
The `await Get`s inside of your loop inhibit parallelism (as each target will be processed sequentially). To allow for parallelism, you'd want to move toward using `MultiGet` for more of this. See #7491 for more info.

Reply:
Yeah, I went back and forth on this, mostly for UX reasons. But now I recall based on a comment in the original setup.py task that at least the setup.py executions themselves must run sequentially, in dep order, because if exportable target B depends on exportable target A, and the setup.py command we run involves publishing (e.g., `sdist upload`), then we don't want to publish B before A.

Maybe the above behavior should be controlled by an option, so that we can get concurrency when the publishing dependency thing isn't an issue. I'll address this one way or another.

Reply:
> and the setup.py command we run involves publishing (e.g., sdist upload), then we don't want to publish B before A.

...oof. That's scary.

A strategy that would still work for that case without being embarassingly parallel would be to do this recursively (setup.py for X depends on setup.py for Y).